Update to magic-sdk 0.17 and refactor

ZYJLiu opened the pull request:
Main changes:
- add UserContext
- add Web3Context
- change network from Goerli to Sepolia
- refactor components to use new custom useContext hooks

ZYJLiu committed:
update magic-sdk

ZYJLiu committed:
refactor

ZYJLiu committed:
refactor

ZYJLiu committed:
remove unused

jamesrp13 commented:
@hcote would you mind glancing through this and letting us know if the way we're handling the rpc provider matches your expectation?

hcote commented on frontend/src/components/AppHeader.tsx:
```diff
@@ -1,53 +1,57 @@
+    try {
+      // Try to show the magic wallet UI
+      // This will only work if the user has connected via a magic wallet, not via browser wallet (e.g. MetaMask)
+      await magic.wallet.showUI();
```
One thing you could do here is catch the case where a user logs out from the settings page that's available through the showUI() method.

The request will emit a `promiEvent` called `disconnect`. Catching it would look like this

```
  magic.wallet.showUI().on('disconnect', () => {
    console.log('user disconnected');
    disconnect()
  });
```

<img width="467" alt="Screenshot 2023-05-23 at 9 09 13 AM" src="https://github.com/magiclabs/magic-nft-template/assets/34493689/40db1632-6117-442b-b71c-2ff9a084095e">

jamesrp13 replied:
@hcote completely missed this, thanks for the callout

hcote commented on frontend/src/components/AppHeader.tsx:
```diff
@@ -1,53 +1,57 @@
+    await magic.user.logout();
+
+    // Clear the user state
+    setUser(null);
```
After a user disconnects, are you re-initializing the web3 instance? Thinking of the case where a user logs in with metamask, then disconnects, the web3 provider will be metamask even though they disconnected - any non-logged in state should have the web3 provider be the magic.rpcProvider (that will be returned by magic.wallet.getProvider() when there's no user logged in)

jamesrp13 replied:
Another good call. Will fix

hcote commented:
I didn't see any handling for events that would be emitted by third party wallets (`chainChanged`, `accountsChanged`) - if a user connects with metamask and then switches their wallet, is this app handling that in real time? Or is the user required to reload the page in order for the app to pull in the new account?

ZYJLiu committed:
implement hunter feedback

jamesrp13 commented:
> I didn't see any handling for events that would be emitted by third party wallets (`chainChanged`, `accountsChanged`) - if a user connects with metamask and then switches their wallet, is this app handling that in real time? Or is the user required to reload the page in order for the app to pull in the new account?

This is an interesting one because it's a UX consideration we hadn't taken into account when we first built this. We only support one chain here since it's using a contract deployed to that chain. Presumably the right way to handle this would be to listen for the chainChanged message and display something in the UI that indicates they aren't on the right chain? And then on the accountsChanged front, would you suggest we log them in with the new account? Do you have any sample code for this? cc @ZYJLiu

ZYJLiu commented:
Implemented Hunter's feedback
- updated `magic.wallet.showUI()`
- reinitialize web3 on after disconnecting
- add `onAccountChange` event listener to provider

jamesrp13 commented on frontend/src/context/UserContext.tsx:
```diff
@@ -54,7 +55,7 @@ export const UserProvider = ({ children }: { children: React.ReactNode }) => {
 
     fetchData();
-  }, [web3]);
+  }, [web3, isAccountChanged]);
```
@ZYJLiu what happens here if you switch to an account that you've never approved to connect before? I imagine we'd have to go through the login process again there

hcote replied:
If a user changes their connected metamask account, you can just update the state where you're storing the user's address to reflect that. In our demo app we just do `setUser(newAccount)`

jamesrp13 commented on frontend/src/components/AppHeader.tsx:
```diff
@@ -20,7 +20,10 @@ export default function AppHeader({}) {
       // Try to show the magic wallet UI
       // This will only work if the user has connected via a magic wallet, not via browser wallet (e.g. MetaMask)
-      await magic.wallet.showUI();
+      /// @ts-ignore
```
2 things:
1. What's the error that we're ignoring with @ts-ignore
2. Is showUI an async method? If so we should await it right?

hcote replied:
1. It's an async function so it's type is `Promise<boolean>` which does not have a `.on()` but really it returns a `PromiEvent` which is a specific type of promise that has a `.on('event')`. The type error can be ignored for now. 
2. It is async but there's no logic waiting to be run afterward it resolves so awaiting/not awaiting shouldn't matter

hcote commented:
> Presumably the right way to handle this would be to listen for the chainChanged message and display something in the UI that indicates they aren't on the right chain? 

Yup, that sounds good to me. Even just an alert should suffice. 

> And then on the accountsChanged front, would you suggest we log them in with the new account? Do you have any sample code for this?

This is the code in our demo app for handling this - feel free to edit for this app as appropriate. Note that you should handle `!acc[0]` because if a user disconnects their metamask, it will emit an `accountsChanged` event but the `accounts` will be `[]`.

```
  // Update state for newly connected wallet
  const handleAccountsChanged = (acc: any) => {
    console.log('New account:', acc);
    // If user disconnected wallet, logout & reset web3
    if (!acc[0]) {
      handleDisconnect();
    } else {
      localStorage.setItem('user', acc[0]);
      setUser(acc[0]);
    }
  };
```

ZYJLiu committed:
add on chainChanged alert

ZYJLiu commented on frontend/src/context/Web3Context.tsx:
```diff
@@ -0,0 +1,95 @@
+      const web3Instance = new Web3(provider);
+
+      // Subscribe to accounts changed event
+      provider.on("accountsChanged", async () => {
+        setIsAccountChanged((state) => !state);
+      });
+
+      // Subscribe to chain changed event
+      provider.on("chainChanged", async () => {
+        const chainId = await web3Instance.eth.getChainId();
+        const sepoliaChainId = 11155111;
+        if (chainId !== sepoliaChainId) {
+          alert("Please switch to the Sepolia network");
+        }
+      });
```
> I didn't see any handling for events that would be emitted by third party wallets (chainChanged, accountsChanged) - if a user connects with metamask and then switches their wallet, is this app handling that in real time? Or is the user required to reload the page in order for the app to pull in the new account?

Updated with chainChanged and accountsChanged. For accountsChanged, it will trigger `fetchData` to run in the UserContext to update the user state.
